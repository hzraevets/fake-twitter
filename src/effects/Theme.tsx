import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  useCallback,
  createContext,
  useState,
  ReactNode,
} from 'react';
import { ConfigProvider, theme as antdTheme } from "antd";

interface ThemeContextInterface {
  theme: string;
  setTheme: Dispatch<SetStateAction<string>>;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextInterface>({
  theme: 'light',
  setTheme: () => null,
  toggleTheme: () => null,
});

const getTheme = () => {
  const theme = localStorage.getItem('theme');
  if (!theme) { // default theme
    localStorage.setItem('theme', 'light');
    return 'light';
  }

  return theme;
};

const ThemeProvider = ({ children }: { children: ReactNode }) => {
  const [ theme, setTheme ] = useState(getTheme);

  function toggleTheme() {
    if (theme === 'dark') {
      setTheme('light');
    } else {
      setTheme('dark');
    }
  }

  const refreshTheme = useCallback(() => {
    localStorage.setItem('theme', theme);
  }, [theme]);

  useEffect(() => {
    refreshTheme();
  }, [refreshTheme, theme]);

  return (
    <ThemeContext.Provider
      value={{
        theme,
        setTheme,
        toggleTheme,
      }}
    >
      <div className={`theme-container ${theme}`}>
        <ConfigProvider
          theme={{
            algorithm: theme === 'dark'
              ? antdTheme.darkAlgorithm
              : antdTheme.defaultAlgorithm
          }}
        >
          {children}
        </ConfigProvider>
      </div>
    </ThemeContext.Provider>
  );
};

export { ThemeContext, ThemeProvider };
