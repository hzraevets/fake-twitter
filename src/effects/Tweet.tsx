import React, { createContext, useState, ReactNode } from 'react';
import getUnixTime from 'date-fns/getUnixTime';

import {
  Tweet,
  TweetStorage,
  CreateNewTweetParams,
  ReadNonDeleteTweetParams,
  ReadNonDeleteTweetResponse,
} from 'models';

export interface TweetContextInterface {
  allTweets: TweetStorage;
  nextTweetId: number;
  addNewTweet: (params: CreateNewTweetParams) => void;
  readTimeline: (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>;
  editTweet: (id: number, content: string) => void;
  deleteTweet: (id: number) => void;
}

const initialTweetStorage: TweetStorage = {
  hashMap: {},
  order: [],
};

const TweetContext = createContext<TweetContextInterface>({
  allTweets: initialTweetStorage,
  nextTweetId: 1,
  addNewTweet: () => null,
  readTimeline: () => Promise.resolve({ data: [] }),
  editTweet: () => null,
  deleteTweet: () => null,
});

const TweetProvider = ({ children }: { children: ReactNode }) => {
  const [ allTweets, setAllTweets ] = useState(initialTweetStorage);
  const [ nextTweetId, setNextTweetId ] = useState(1);

  function addNewTweet(params: CreateNewTweetParams) {
    const now = getUnixTime(new Date());
    const { hashMap, order } = allTweets;

    const newTweet: Tweet = {
      ...params,
      id: nextTweetId,
      createTime: now,
      updateTime: now,
      isDeleted: false,
    };

    setNextTweetId(nextTweetId + 1);
    setAllTweets({
      hashMap: { ...hashMap, [nextTweetId]: newTweet },
      order: [ nextTweetId, ...order ],
    });
  }

  function readTimeline({ start, offset = 1 }: ReadNonDeleteTweetParams) {
    const data: Tweet[] = [];
    let nextCursor: number | null;

    const { hashMap, order } = allTweets;

    for (nextCursor = start; (data.length < offset) && nextCursor < order.length; nextCursor++) {
      const tweet = hashMap[order[nextCursor]];

      if (!tweet.isDeleted) {
        data.push(tweet);
      }
    }

    const res: ReadNonDeleteTweetResponse = { data };

    if (nextCursor < order.length) {
      res.nextCursor = nextCursor;
    }

    return new Promise<ReadNonDeleteTweetResponse>(resolve =>
      setTimeout(() => resolve(res), Math.random() * 1000));
  }

  function editTweet(id: number, content: string) {
    const { hashMap, order } = allTweets;
    const oldTweet = hashMap[id];

    if (!oldTweet) {
      // This should not happen!
      return;
    }

    hashMap[id].content = content;
    hashMap[id].updateTime = getUnixTime(new Date());

    setAllTweets({ hashMap, order });
  }

  function deleteTweet(id: number) {
    const { hashMap } = allTweets;
    let order = allTweets.order;
    const oldTweet = hashMap[id];

    if (!oldTweet) {
      // This should not happen!
      return;
    }

    hashMap[id].isDeleted = true;
    order = order.filter(tid => tid !== id);

    setAllTweets({ hashMap, order });
  }

  return (
    <TweetContext.Provider
      value={{
        allTweets,
        nextTweetId,
        addNewTweet,
        readTimeline,
        editTweet,
        deleteTweet,
      }}
    >
      {children}
    </TweetContext.Provider>
  );
}

export { TweetContext, TweetProvider };
