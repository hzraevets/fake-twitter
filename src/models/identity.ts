export interface RegisterField {
  username: string;
  firstname: string;
  password: string;
}

export type LoginField = Omit<RegisterField, 'firstname'>

export type User = RegisterField;
