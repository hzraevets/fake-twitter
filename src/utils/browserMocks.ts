import '@testing-library/jest-dom';

// https://stackoverflow.com/questions/65282181/how-to-use-jest-for-testing-a-react-component-with-localstorage
const localStorageMock = (function() {
  let store: {[key: string]: string} = {};

  return {
    getItem: function(key: string) {
      return store[key] || null;
    },
    setItem: function(key: string, value: any) {
      store[key] = String(value);
    },
    removeItem: function(key: string) {
      delete store[key];
    },
    clear: function() {
      store = {};
    }
  }
})();

export const enableLocalStorageMock = () => Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});

// https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
export const enableMatchMedia = () => Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
