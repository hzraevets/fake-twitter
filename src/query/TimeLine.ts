import { QueryClient } from 'react-query';

import { ReadNonDeleteTweetParams, ReadNonDeleteTweetResponse } from 'models';

export const timeLineQuery = new QueryClient();

export const generateFetchTimeline = (
  readTimeline:  (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>
) => ({ pageParam = 0 }) => {
  return readTimeline({ start: pageParam });
}

export const clearThenReFetch = async (
  readTimeline:  (params: ReadNonDeleteTweetParams) => Promise<ReadNonDeleteTweetResponse>
) => {
  timeLineQuery.clear();

  await setTimeout(() => null, 500);

  return timeLineQuery.prefetchInfiniteQuery('timeline', generateFetchTimeline(readTimeline), {
    getNextPageParam: (lastPage, pages) => {
      return lastPage.nextCursor;
    },
  });
}
