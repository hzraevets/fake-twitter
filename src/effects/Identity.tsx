import React, {
  Dispatch,
  SetStateAction,
  useEffect,
  createContext,
  useState,
  ReactNode,
} from 'react';

import { UserStorage, User } from 'models';
import { md5 } from 'utils/md5';

export interface IdentityContextInterface {
  identity: UserStorage;
  setIdentity: Dispatch<SetStateAction<UserStorage>>;
  loginUser: string | null;
  setLoginUser: Dispatch<SetStateAction<string | null>>;
  logout: () => void;
  register: (newUser: User) => UserStorage;
}

const IdentityContext = createContext<IdentityContextInterface>({
  identity: {},
  setIdentity: () => null,
  loginUser: null,
  setLoginUser: () => null,
  logout: () => null,
  register: () => ({}),
});

const getIdentity = (): UserStorage => {
  const identity = localStorage.getItem('identity');
  if (!identity) {
    localStorage.setItem('identity', '{}');
    return {};
  }

  return JSON.parse(identity);
};

const getLoginUser = (): string | null => {
  const loginUser = localStorage.getItem('loginUser');
  if (!loginUser) {
    return null;
  }

  return loginUser;
};

const IdentityProvider = ({ children }: { children: ReactNode }) => {
  const [ identity, setIdentity ] = useState(getIdentity);
  const [ loginUser, setLoginUser ] = useState(getLoginUser);

  function logout() {
    if (loginUser) {
      setLoginUser(null);
      localStorage.removeItem('loginUser');
    }
  }

  function register(newUser: User) {
    const newStorage: UserStorage = {
      ...identity,
      [newUser.username]: {
        ...newUser,
        password: md5(newUser.password),
      },
    };

    localStorage.setItem('identity', JSON.stringify(newStorage));
    return newStorage;
  }

  useEffect(() => {
    const updateIdentity = () => {
      localStorage.setItem('identity', JSON.stringify(identity));
    };

    updateIdentity();
  }, [identity]);

  useEffect(() => {
    const executeLogin = () => {
      if (loginUser) {
        localStorage.setItem('loginUser', loginUser);
      }
    };

    executeLogin();
  }, [loginUser]);

  return (
    <IdentityContext.Provider
      value={{
        identity,
        setIdentity,
        loginUser,
        setLoginUser,
        logout,
        register,
      }}
    >
      {children}
    </IdentityContext.Provider>
  );
};

export { IdentityContext, IdentityProvider };
