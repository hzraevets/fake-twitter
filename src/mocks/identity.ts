import { User, UserStorage } from 'models';
import { md5 } from 'utils/md5';

export const testUser: User = {
  username: 'testUser',
  firstname: 'Test',
  password: 'abcde',
}

export const liBai: User = {
  username: 'liBai',
  firstname: 'Li',
  password: '12345',
};

export const storage: UserStorage = {
  testUser: { ...testUser, password: md5(testUser.password) },
  liBai: { ...liBai, password: md5(liBai.password) },
}
