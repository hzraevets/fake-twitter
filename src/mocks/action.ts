import { enableLocalStorageMock, enableMatchMedia } from 'utils/browserMocks';
import { TweetContextInterface } from 'effects/Tweet';
import {ReadNonDeleteTweetParams, ReadNonDeleteTweetResponse, Tweet} from 'models';
import { storage as identityStorage, testUser } from './identity';
import { nextTweetId, storage as tweetStorage } from './tweet';

export function prepareTweetProviderPropsWithoutCRUD(): TweetContextInterface {
  enableMatchMedia();
  enableLocalStorageMock();

  // prepare work for identity
  localStorage.setItem('identity', JSON.stringify(identityStorage));
  localStorage.setItem('loginUser', testUser.username);

  // prepare tweet provider props and return
  return {
    allTweets: tweetStorage,
    nextTweetId,
    addNewTweet: () => null,
    readTimeline: () => Promise.resolve({ data: [] }),
    editTweet: () => null,
    deleteTweet: () => null,
  };
}

export function prepareTweetProviderPropsWithRead(): TweetContextInterface {
  enableMatchMedia();
  enableLocalStorageMock();

  // prepare work for identity
  localStorage.setItem('identity', JSON.stringify(identityStorage));
  localStorage.setItem('loginUser', testUser.username);

  // prepare tweet provider props and return
  return {
    allTweets: tweetStorage,
    nextTweetId,
    addNewTweet: () => null,
    readTimeline: ({ start, offset = 1 }: ReadNonDeleteTweetParams) => {
      const dat: Tweet[] = [];
      let nextCursor: number | null;

      const { hashMap, order } = tweetStorage;

      for (nextCursor = start; (dat.length < offset) && nextCursor < order.length; nextCursor++) {
        const tweet = hashMap[order[nextCursor]];

        if (!tweet.isDeleted) {
          dat.push(tweet);
        }
      }

      const res: ReadNonDeleteTweetResponse = { data: dat };

      if (nextCursor < order.length) {
        res.nextCursor = nextCursor;
      }

      return Promise.resolve<ReadNonDeleteTweetResponse>(res);
    },
    editTweet: () => null,
    deleteTweet: () => null,
  };
}
