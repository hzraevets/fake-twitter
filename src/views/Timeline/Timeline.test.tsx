import React from 'react';
import { act } from 'react-dom/test-utils';
import { render } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { QueryClientProvider } from 'react-query';
// @ts-ignore
import { create } from 'react-test-renderer';

import { Timeline } from './';
import { enableLocalStorageMock, enableMatchMedia } from 'utils/browserMocks';
import { IdentityProvider } from 'effects/Identity';
import { timeLineQuery } from 'query/TimeLine';
import { storage, testUser } from 'mocks/identity';

describe('Timeline', () => {
  const { username } = testUser;

  beforeEach(() => {
    enableMatchMedia();
    enableLocalStorageMock();

    // prepare work for identity
    localStorage.setItem('identity', JSON.stringify(storage));
    localStorage.setItem('loginUser', testUser.username);
  });

  it('render Timeline page', async () => {

    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      // https://reactrouter.com/en/6.8.1/router-components/memory-router
      render((
        <MemoryRouter initialEntries={[`/${username}`]}>
          <Routes>
            <Route
              path=":username"
              element={
                <QueryClientProvider client={timeLineQuery}>
                  <IdentityProvider>
                    <Timeline/>
                  </IdentityProvider>
                </QueryClientProvider>
              }
            />
          </Routes>
        </MemoryRouter>
      ));
    });
  });

  it('render Timeline to snapshot', async () => {
    const renderer = await act(async () => {
      return create(
        <MemoryRouter initialEntries={[`/${username}`]}>
          <Routes>
            <Route
              path=":username"
              element={
                <QueryClientProvider client={timeLineQuery}>
                  <IdentityProvider>
                    <Timeline/>
                  </IdentityProvider>
                </QueryClientProvider>
              }
            />
          </Routes>
        </MemoryRouter>
      );
    });
    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
