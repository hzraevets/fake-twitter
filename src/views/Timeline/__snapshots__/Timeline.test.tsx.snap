// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline render Timeline to snapshot 1`] = `
<div
  className="timeline-container flex h-full overflow-auto"
  id="scrollableDiv"
>
  <div
    className="timeline relative pt-10 mx-auto gap-4 columns-1 md:columns-2"
  >
    <div
      className="break-inside-avoid-column"
    >
      <div
        className="ant-card ant-card-bordered w-80 break-inside-avoid-column mb-4 css-dev-only-do-not-override-diro6f"
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="timeline-user-bg"
            className="bg-slate-500 !hidden md:!block"
            src="logo.svg"
          />
        </div>
        <div
          className="ant-card-body"
          style={Object {}}
        >
          <div
            className="absolute top-2 right-2 cursor-pointer select-none"
            onClick={[Function]}
          >
            <button
              aria-checked={false}
              className="ant-switch css-dev-only-do-not-override-diro6f"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="switch"
              type="button"
            >
              <div
                className="ant-switch-handle"
              />
              <span
                className="ant-switch-inner"
              >
                <span
                  className="ant-switch-inner-checked"
                />
                <span
                  className="ant-switch-inner-unchecked"
                />
              </span>
            </button>
          </div>
          <div
            className="ant-card-meta"
          >
            <div
              className="ant-card-meta-avatar"
            >
              <span
                className="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-icon bg-slate-500 css-dev-only-do-not-override-diro6f"
                style={Object {}}
              >
                <span
                  aria-label="user"
                  className="anticon anticon-user"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="user"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                    />
                  </svg>
                </span>
              </span>
            </div>
            <div
              className="ant-card-meta-detail"
            >
              <div
                className="ant-card-meta-title"
              >
                <div
                  className="text-2xl font-bold leading-none"
                >
                  Test
                </div>
                <div
                  className="text-xs font-medium text-slate-500"
                >
                  @
                  testUser
                </div>
              </div>
            </div>
          </div>
        </div>
        <ul
          className="ant-card-actions"
        >
          <li
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <span>
              <span
                className="hover:text-red-600 select-none inline-block w-full"
                onClick={[Function]}
              >
                Logout
              </span>
            </span>
          </li>
        </ul>
      </div>
      <div
        className="ant-card ant-card-bordered w-80 break-inside-avoid-column mb-4 css-dev-only-do-not-override-diro6f"
      >
        <div
          className="ant-card-body"
          style={Object {}}
        >
          <div
            className="ant-space css-dev-only-do-not-override-diro6f ant-space-vertical w-full"
            style={Object {}}
          >
            <div
              className="ant-space-item"
              style={
                Object {
                  "marginBottom": 16,
                }
              }
            >
              <textarea
                className="ant-input max-h-32 css-dev-only-do-not-override-diro6f"
                disabled={false}
                onChange={[Function]}
                onCompositionEnd={[Function]}
                onCompositionStart={[Function]}
                onKeyDown={[Function]}
                placeholder="What's happening?"
                rows={3}
                style={
                  Object {
                    "resize": undefined,
                  }
                }
                value=""
              />
            </div>
            <div
              className="ant-space-item"
              style={Object {}}
            >
              <button
                className="ant-btn css-dev-only-do-not-override-diro6f ant-btn-primary float-right"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                <span>
                  Tweet
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-card ant-card-bordered w-80 overflow-auto break-inside-avoid-column mb-10 css-dev-only-do-not-override-diro6f"
    >
      <div
        className="ant-card-body"
        style={Object {}}
      >
        <div
          className="infinite-scroll-component__outerdiv"
          style={Object {}}
        >
          <div
            className="infinite-scroll-component "
            style={
              Object {
                "WebkitOverflowScrolling": "touch",
                "height": "auto",
                "overflow": "auto",
              }
            }
          >
            <div
              className="ant-divider css-dev-only-do-not-override-diro6f ant-divider-horizontal ant-divider-with-text ant-divider-with-text-center ant-divider-plain select-none"
              role="separator"
            >
              <span
                className="ant-divider-inner-text"
                style={Object {}}
              >
                It is all, nothing more ü§ê
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
